anvil/plot.py

```diff
@@ -0,0 +1,193 @@
+import numpy as np
+import matplotlib.pyplot as plt
+from matplotlib.ticker import MaxNLocator
+import torch
+
+from reportengine.figure import figure
+
+from anvil.observables import bootstrap
+
+@figure
+def plot_zero_momentum_2pf(zero_momentum_2pf, training_geometry):
+    error = bootstrap(zero_momentum_2pf)
+    fig, ax = plt.subplots()
+    ax.errorbar(
+        x=range(training_geometry.length),
+        y=zero_momentum_2pf[0],
+        yerr=error,
+        fmt="-r",
+        label=f"L = {training_geometry.length}",
+    )
+    ax.set_yscale("log")
+    ax.set_ylabel("$\hat{G}(0, t)$")
+    ax.set_xlabel("$t$")
+    ax.set_title("Zero momentum two point function")
+    return fig
+
+
+@figure
+def plot_effective_pole_mass(training_geometry, effective_pole_mass):
+    error = bootstrap(effective_pole_mass)
+    fig, ax = plt.subplots()
+    ax.errorbar(
+        x=range(1, training_geometry.length - 1),
+        y=effective_pole_mass[0],
+        yerr=error,
+        fmt="-b",
+        label=f"L = {training_geometry.length}",
+    )
+    ax.set_ylabel("$m_p^{eff}$")
+    ax.set_xlabel("$t$")
+    ax.set_title("Effective pole mass")
+    return fig
+
+
+@figure
+def plot_2pf(training_geometry, two_point_function):
+    corr = np.empty((training_geometry.length, training_geometry.length))
+    std = np.empty((training_geometry.length, training_geometry.length))
+    for t in range(training_geometry.length):
+        for x in range(training_geometry.length):
+            corr[x, t] = float(two_point_function(t, x)[0])
+            std[x, t] = float(bootstrap(two_point_function(t, x)))
+
+    fractional_std = std / corr
+
+    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(13, 6), sharey=True)
+    ax2.set_title(r"$\sigma_G / G$")
+    ax1.set_title("$G(x, t)$")
+    ax1.set_xlabel("$x$")
+    ax2.set_xlabel("$x$")
+    ax1.set_ylabel("$t$")
+    im1 = ax1.pcolor(corr)
+    im2 = ax2.pcolor(fractional_std)
+
+    ax1.yaxis.set_major_locator(MaxNLocator(integer=True))
+    ax1.xaxis.set_major_locator(MaxNLocator(integer=True))
+    ax2.xaxis.set_major_locator(MaxNLocator(integer=True))
+
+    fig.colorbar(im1, ax=ax1)
+    fig.colorbar(im2, ax=ax2)
+
+    return fig
+
+
+@figure
+def plot_volume_averaged_2pf(volume_averaged_2pf):
+    fig, ax = plt.subplots()
+    ax.set_title("Volume-averaged two point function")
+    ax.set_ylabel("$G_k(0,0)$")
+    ax.set_xlabel("$t$")
+    ax.plot(volume_averaged_2pf(0, 0), "-")
+    return fig
+
+
+@figure
+def plot_autocorrelation_2pf(autocorrelation_2pf):
```

Comment:
do we need 4 subplots here, why not just make 4 seperate figures? Personally I would make autocorrelation 4 seperate functions which depend on each other and just return one thing not a tuple. Then I would make a seperate plot for each one, perhaps I'm missing something though